File: Problems/Algorithms/146. LRU Cache/lru_cache.py
#!/usr/bin/env python3


class Node:
    def __init__(self, key: int, value: int):
        self.key, self.value = key, value
        self.prev, self.next = None, None


class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = dict()
        
        self._sentinel = Node(-1, -1)
        self._sentinel.next = self._sentinel.prev = self._sentinel

    def _remove(self, node: Node):
        prev, next = node.prev, node.next
        prev.next = next
        next.prev = prev

    def _insert(self, node: Node):
        prev = self._sentinel.prev
        prev.next = node
        node.next = self._sentinel
        node.prev = prev
        self._sentinel.prev = node

    def get(self, key: int) -> int:
        if key in self.cache:
            node = self.cache[key]
            self._remove(node)
            self._insert(node)
            return node.value
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self._remove(self.cache[key])
        node = Node(key, value)
        self._insert(node)
        self.cache[key] = node

        if len(self.cache) > self.capacity:
            node = self._sentinel.next
            self._remove(node)
            del self.cache[node.key]


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: Problems/Algorithms/146. LRU Cache/lru-cache.cpp
#include <unordered_map>

using namespace std;

class LRUCache {
public:
    class Node {
        public:
            int key;
            int val;
            Node* prev;
            Node* next;

            Node(int key, int val) {
                this->key = key;
                this->val = val;
            }
    };

    Node* head = new Node(-1, -1);
    Node* tail = new Node(-1, -1);

    int capacity;
    unordered_map<int, Node*> map;

    LRUCache(int capacity) {
        this->capacity = capacity;
        head->next = tail;
        tail->prev = head;
    }

    void add(Node* node) {
        Node* temp = head->next;

        node->next = temp;
        node->prev = head;

        head->next = node;
        temp->prev = node;
    }

    void remove(Node* node) {
        Node* tempPrev = node->prev;
        Node* tempNext = node->next;

        tempPrev->next = tempNext;
        tempNext->prev = tempPrev;
    }
    
    int get(int key) {
        if (map.find(key) != map.end()) {
            Node* result = map[key];
            int ans = result->val;

            map.erase(key);
            remove(result);
            add(result);

            map[key] = head->next;
            return ans;
        }

        return -1;
    }
    
    void put(int key, int value) {
        if (map.find(key) != map.end()) {
            Node* curr = map[key];
            map.erase(key);
            remove(curr);
        }

        if (map.size() == capacity) {
            map.erase(tail->prev->key);
            remove(tail->prev);
        }

        add(new Node(key, value));
        map[key] = head->next;
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */